### PR TITLE
Peerset cleanup

### LINKED ISSUE #5642: Peerset: The priority groups's relationship with reserved-only is kind of blurry and buggy
In reserved-only mode, we actively establish connections only to reserved nodes:
https://github.com/paritytech/substrate/blob/8991aab9d98da49e3a7e608cb0a0162b610e25c7/client/peerset/src/lib.rs#L379-L383

But in reserved-only mode, we also accept incoming connections from nodes, whatever their priority group is:
https://github.com/paritytech/substrate/blob/8991aab9d98da49e3a7e608cb0a0162b610e25c7/client/peerset/src/peersstate.rs#L266

Setting the reserved-only mode to true also disconnects all non-reserved nodes, even if they were in a different priority group:
https://github.com/paritytech/substrate/blob/8991aab9d98da49e3a7e608cb0a0162b610e25c7/client/peerset/src/lib.rs#L264

This is all confused.

### PR DESCRIPTION
**Pull request by tomaka**
Fix #5642
As detailed in #5642, the current code is a bit of a mess.

This PR removes `PeersState::is_priority` and replaces `PeersState::priority_nodes` with `no_slot_nodes`. This restores the `PeersState` struct to how it was originally designed, so that its only responsibility is to handle reputations and slots.

Moves out `priority_nodes` and `reserved_nodes` to the `Peerset` itself, which consults these lists to determine who to connect to.
The priority groups and reserved nodes are now two separate things, and no longer in a single hashmap with a special key.

I slightly changed the behaviour: nodes in `reserved_nodes` do not occupy slots, but nodes in `priority_groups` do. Nodes in `priority_groups` have a higher priority for outgoing slots compared to non-priority nodes.

As a consequence, sentry nodes passed with the `--sentry-nodes` CLI option and validator nodes passed with the `--sentry` CLI option are now also put in `reserved_nodes`.

Note that the point of this PR is to fix #5642, **not** to discuss the design of priority groups. I can change the priority nodes back to not occupy slots if that's demanded, but I'm aware that there are lots of ambiguities about them. Fixing that is out of scope of this PR.

### CODE REVIEW

**Comment by arkpar** [comment]
> I slightly changed the behaviour: nodes in reserved_nodes do not occupy slots, but nodes in priority_groups do. 

Any specific reason for that? Priority groups are supposed to ignore slot limits. If e.g. a validator needs to connect to specific collators it should do so regardless of peers limit specified on the CLI.

> Note that the point of this PR is to fix #5642, not to discuss the design of priority groups.

Well, it does change the design of priority groups so it's hard to review without discussing it.

> I can change the priority nodes back to not occupy slots if that's demanded, but I'm aware that there are lots of ambiguities about them.

Such as?

File: client/peerset/src/lib.rs [hunk 1]
```diff
@@ -527,11 +603,6 @@ impl Peerset {
 	pub fn num_discovered_peers(&self) -> usize {
 		self.data.peers().len()
 	}
-
-	/// Returns priority group by id.
-	pub fn get_priority_group(&self, group_id: &str) -> Option<HashSet<PeerId>> {
```
Comment (arkpar) [inline client/peerset/src/lib.rs:532]: This should be exposed to RPC really. I'd keep it, but rename to `priority_group`

**Comment by arkpar** [comment]
Looks good, except priority group peers should not occupy slots.

**Comment by tomaka** [comment]
> Any specific reason for that? Priority groups are supposed to ignore slot limits. If e.g. a validator needs to connect to specific collators it should do so regardless of peers limit specified on the CLI.

> Such as?

Right now they're used for two things:

- `authority-discovery` (provided it's working) adds all sentry nodes of all validators to a priority group.
- We add the sentries and validators passed with `--sentry` or `--sentry-nodes` to another group.

For the first one, clearly we don't want to be connected to potentially all 200-250 other validators (and more in the future). For the second one, however, we do.

**Comment by arkpar** [comment]
That's why it's up to the API user to manage the group. `authority-discovery` should be setting sensible number of peers, based on the criteria that is specific to that group. Delegating it to the peerset does not seem right.

**Comment by mxinden** [comment]
> authority-discovery (provided it's working, which is not really the case yet) adds all sentry nodes of all validators to a priority group.

@tomaka @arkpar authority discovery is not adding all sentry nodes of all validators to the priority group, but only a stable subset [currently set to 10](https://github.com/paritytech/substrate/blob/master/client/authority-discovery/src/addr_cache.rs#L29).

This is also exposed via the [authority_discovery_priority_group_size](https://github.com/paritytech/substrate/blob/master/client/authority-discovery/src/lib.rs#L659) metric, last time I checked the value was `10` for most of our nodes.

> (provided it's working, which is not really the case yet)

Is there something I am missing apart from the validators behind sentry nodes have difficulties querying the DHT?

**Comment by tomaka** [comment]
Does that then mean that `authority-discovery` is supposed to track who we are connected to, and have some sort of timeout system, and remove from the set the nodes that we seemingly can't reach?

Since the peerset is responsible for determining who we are connected to, it makes way more sense to me to indicate to it a list of nodes that we would like to be connected to, and let it handle the rest.

> Is there something I am missing apart from the validators behind sentry nodes have difficulties querying the DHT?

I'm just judging based on the high number of failures to publish and query records, and on the fact that nodes only have 1.5 records in their local storage on average.

**Comment by tomaka** [comment]
Another ambiguity that should be clarified is: why do we still have the concept of "reserved peers", assuming that priority groups consume slots?

**Comment by arkpar** [comment]
> Does that then mean that authority-discovery is supposed to track who we are connected to, and have some sort of timeout system, and remove from the set the nodes that we seemingly can't reach?

> Since the peerset is responsible for determining who we are connected to, it makes way more sense to me to indicate to it a list of nodes that we would like to be connected to, and let it handle the rest.

That's a good point. In that case I suppose priority group API should accept an ordered list of peers along with a max number of additional slots. Peerset would prioritize based on specified order or reputation.  Currently it just tries to connect the priority in unspecified order, ignoring reputation which may lead to all basic slots taken by peers with bad reputation. 

There's also a problem of priority groups competing for slots currently. What if total number of slots is less than all peers in all priority groups? How would you pick which get to connect? With this PR it's random and unfair.

> Another ambiguity that should be clarified is: why do we still have the concept of "reserved peers", assuming that priority groups don't use slots?

My understanding is that reserved peers is simply a priority group that user may edit explicitly. 
Additionally `reserved-only` mode should indeed only allow connections to peers from that group, ignoring other priority groups.

**Comment by tomaka** [comment]
> Note that the point of this PR is to fix #5642, not to discuss the design of priority groups. I can change the priority nodes back to not occupy slots if that's demanded, but I'm aware that there are lots of ambiguities about them. Fixing that is out of scope of this PR.

I wrote this small line in order to prevent the discussion from going off-topic.

I'm going to change priority groups to occupy slots again, and we'll see later for further changes.

**Comment by tomaka** [comment]
Should be ready.

**Comment by arkpar** [comment]
As I've mentioned before, `reserved-only` should only allow connections to reserved peers, disregarding other priority groups. This is a hard filter to allow user to limit all communication to a controlled set.

File: client/peerset/src/peersstate.rs [hunk 2]
```diff
@@ -212,170 +193,40 @@ impl PeersState {
...
-	/// Note that reserved nodes don't count towards the number of slots.
-	fn try_accept_incoming(&mut self, peer_id: &PeerId) -> bool {
-		let is_priority = self.is_priority(peer_id);
-
-		// We are only accepting connections from priority nodes.
-		if !is_priority && self.priority_only {
-			return false;
+	/// Has no effect if the peer was not in the group.
+	pub fn remove_no_slot_node(&mut self, peer_id: &PeerId) {
+		// Reminder: `HashSet::remove` returns falsetrue if the node was in the set
```
Comment (andresilva) [inline client/peerset/src/peersstate.rs:218]:
```suggestion
		// Reminder: `HashSet::remove` returns false if the node was already not in the set
```

File: client/peerset/src/peersstate.rs [hunk 3]
```diff
@@ -520,51 +395,84 @@ impl<'a> NotConnectedPeer<'a> {
...
+		}
+
+		if let Some(mut peer) = self.state.nodes.get_mut(&*self.peer_id) {
+			peer.connection_state = ConnectionState::Out;
+			if !is_no_slot_occupy {
+				self.state.num_out += 1;
+			}
 		} else {
-			Err(self)
+			debug_assert!(false, "State inconsistency: try_accept_incoming on an unknown node");
```
Comment (andresilva) [inline client/peerset/src/peersstate.rs:414]:
```suggestion
			debug_assert!(false, "State inconsistency: try_outgoing on an unknown node");
```

File: client/peerset/src/lib.rs [hunk 4]
```diff
@@ -235,61 +240,92 @@ impl Peerset {
...
 	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		self.reserved_only = reserved_only;
-		self.data.set_priority_only(reserved_only);
 
 		if self.reserved_only {
-			// Disconnect non-reserved nodes.
-			let reserved = self.data.get_priority_group(RESERVED_NODES).unwrap_or_default();
+			// Disconnect all the nodes that aren't reserved.
 			for peer_id in self.data.connected_peers().cloned().collect::<Vec<_>>().into_iter() {
+				if self.priority_groups.get(RESERVED_NODES).into_iter().flatten().any(|p| *p == peer_id) {
```
Comment (arkpar) [inline client/peerset/src/lib.rs:256]:
```suggestion
				if self.priority_groups.get(RESERVED_NODES).map_or(false, |g| g.contains(&peer_id)) {
```

File: client/peerset/src/lib.rs [hunk 5]
```diff
@@ -235,61 +240,92 @@ impl Peerset {
...
 		}
 	}
 
 	fn on_set_priority_group(&mut self, group_id: &str, peers: HashSet<PeerId>) {
-		self.data.set_priority_group(group_id, peers);
-		self.alloc_slots();
+		// Determine the difference between the current group and the new list.
+		let (to_insert, to_remove) = {
+			let current_group = self.priority_groups.entry(group_id.to_owned()).or_default();
+			let to_insert = peers.iter().filter(|peer| !current_group.iter().any(|p| p == *peer))
```
Comment (arkpar) [inline client/peerset/src/lib.rs:275]:
```suggestion
			let to_insert = peers.iter().filter(|peer| !current_group.contains(*peer))
```

File: client/peerset/src/lib.rs [hunk 6]
```diff
@@ -235,61 +240,92 @@ impl Peerset {
...
+				.cloned().collect::<Vec<_>>();
+			let to_remove = current_group.iter().filter(|peer| !peers.iter().any(|p| p == *peer))
```
Comment (arkpar) [inline client/peerset/src/lib.rs:277]:
```suggestion
			let to_remove = current_group.iter().filter(|peer| !peers.contains(*peer))
```
`HashSet::difference` could also work here

**Pull request merged**